### DISCUSSION TITLE
Потеря сортировки при использовании Playlist.saveWithUpdate

### DISCUSSION THREAD

**Discussion by boredwz** [Chimildic/goofy#108] [Помощь с ошибкой]
Привет, делюсь кодом публичного варианта _Liked_
Функция [SaveWithUpdate](https://chimildic.github.io/goofy/#/func?id=savewithupdate) не подошла, потому что 
> Теряется сортировка заданная в массиве

Ну и конечно хотелось поиграться и сделать что-нибудь свое =)

```
function updateLikedPlaylist() {
  //---------------------------------------------------------------------------
  let likedplaylist = Source.getTracks([{id: 'id плейлиста' },]);
  let trackstoadd = Source.getSavedTracks();
  let trackstodel = Selector.sliceCopy(likedplaylist);
  let spliked = Selector.sliceCopy(trackstoadd);
  Filter.removeTracks(trackstoadd, likedplaylist);
  Filter.removeTracks(trackstodel, spliked);
  //---------------------------------------------------------------------------
  if (trackstoadd.length > 0) {Playlist.saveWithAppend({id: 'id плейлиста', tracks: trackstoadd,});}
  if (trackstodel.length > 0) {Playlist.removeTracks('id плейлиста', trackstodel);}
}
```

**Comment by Chimildic** [comment]
Как данный вариант сохраняет сортировку? `saveWithUpdate` также использует `append`.

Под потерей сортировки подразумевается, что новый порядок (перемешивание старых **и** новых треков) невозможно сделать сохраняя дату добавления. Сортировка только новых треков не затрагивается.

**Reply by boredwz** [reply]
Значит, по поводу замены треков, сохраняя дату добавления - это я понял, оно по логике не должно сохранять старую дату.
Теперь про плейлисты `saveWithUpdate` и вручную через `append`.
Прошел день и появились проблемы сортировки и у моего варианта.

Где-то после ~20-30 срабатываний функции какой-то массив из 50 треков поднимается на верх плейлиста, и этим сбивает сортировку по дате, хотя не должно бы.
![image](https://user-images.githubusercontent.com/87862400/126816547-4b130250-2726-4aae-a15b-b83e0d6a1d3c.png)
![image](https://user-images.githubusercontent.com/87862400/126816617-2fde53dc-0e93-40ea-9b2e-34d6b2d95b34.png)


Так же и с `saveWithUpdate`, но тут массив из 100 треков кидает вверх.
```
function test() {
  let test = Source.getSavedTracks();
  Playlist.saveWithUpdate({id: '49J3H0ByINj3mn0WcuLXSy', tracks: test,});
}
```
![image](https://user-images.githubusercontent.com/87862400/126817313-b5f71db7-5153-48e2-90ce-95c10287cbdd.png)

В чем причина?
Чтобы повторить все как в `Liked` приходится сначала удалять содержимое плейлиста, а потом перезаписывать, даже если изменений нет =(

**Reply by Chimildic** [reply]
По умолчанию [saveWithAppend](https://chimildic.github.io/goofy/#/func?id=savewithappend) добавляет треки в начало. Чтобы было в конец, установите аргумент `toEnd` в значение `true`. В документации есть примеры.

Если вам не нужно сохранять дату добавления, используйте [saveWithReplace](https://chimildic.github.io/goofy/#/func?id=savewithreplace) передавая все лайки без вычислений.

**Reply by boredwz** [reply]
После 5 дней так и ничего корректно не заработало, кроме `saveWithReplace`.
`saveWithAppend`, `saveWithUpdate`, с атрибутом `toEnd: true` и без.... Все та же проблема:
> Где-то после ~20-30 срабатываний функции какой-то массив из 50 треков поднимается на верх плейлиста, и этим сбивает сортировку по дате, хотя не должно бы.

Рабочий код
```
function Liked2Playlist() {
  let trackstoadd = Source.getSavedTracks();
  Playlist.saveWithReplace({id: 'id плейлиста', tracks: trackstoadd,});
}
```

**Reply by Chimildic** [reply]
Проблема не воспроизводится. Независимо от количества запусков, все новые лайки добавляются в конец. Новой сортировки не появляется. 
```js
function debug() {
  let tracks = Source.getSavedTracks();
  Playlist.saveWithUpdate({
    id: '',
    tracks: tracks,
    toEnd: true,
  });
}
```
